packages/headless/src/features/commerce/context/cart/cart-slice.ts

```diff
@@ -0,0 +1,49 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {CartState, getCartInitialState} from './cart-state';
+import {addCartItem, removeCartItem, setCart, updateCartItemQuantity} from './cart-actions';
+import {ProductParam} from '../../../../api/commerce/commerce-api-params';
+
+const generateId = (product: ProductParam) => [
+  product.groupId,
+  product.productId,
+  product.sku
+].filter(Boolean).join('_')
+
+export const cartReducer = createReducer(
+  getCartInitialState(),
+
+  (builder) => {
+    builder
+      .addCase(setCart, (state, {payload}) => {
+        const {cart, cartItems} = payload.cart.reduce((acc, item) => {
+          const id = generateId(item.product)
```

Comment:
Should there be any verification on clashing generated IDs ?

Reply:
What would you think of exposing a `setItem` and `removeItem` actions instead of actions for adding, removing and updating the quantity of an item? This would front the clashing to clients. 

Plus, cart items are now indexed by their productId.

Reply:
> This would front the clashing to clients.

What do you mean by this ? I'm not sure I understand

Reply:
The `setItem` and `removeItem` method names expose that the controller only sets and removes items.

The cart is effectively a map of product id -> quantity. By exposing only `setItem` and `removeItem` methods on the controller, the cart controller becomes dumber. This puts the responsibility of dealing with potential item clashing on the caller instead of on the callee (cart controller).

---
A second option is to deal with clashes:
- Expose `addItem(item: CartItem)`, `removeItem(item: CartItem)` and `updateItemQuantity(productId: string, quantity: number)` methods
- In `addItem`, if a clash is found, make the total quantity the current quantity + new item quantity. Otherwise, add the item normally.
- In `removeItem`, remove the item if it's present, otherwise be silent
- In `updateItemQuantity`, update the item quantity if it's present, otherwise be silent

Reply:
> This puts the responsibility of dealing with potential item clashing on the caller instead of on the callee (cart controller).

If by the caller you mean the implementer, this is not a good idea. 

If there is a clash, we just need to inform implementers, and we don't need to do anything fancy.


> * In addItem, if a clash is found, make the total quantity the current quantity + new item quantity. Otherwise, add > > the item normally.
> * In removeItem, remove the item if it's present, otherwise be silent
> * In updateItemQuantity, update the item quantity if it's present, otherwise be silent

For all 3 scenarios, just a no-op/return early + log an informative warning message (not silent) should do the trick.

Reply:
I've applied your suggestions, without the logging. I didn't log since other slices do not log upon edge cases.

I could throw an Error, but it doesn't seem quite right since these are edge cases and not problematic inputs.